DIPs/4NNN-WGB.md

```diff
@@ -0,0 +1,161 @@
+between a format specifier and its corresponding argument are common errors. By
+embedding the argument in the format string tends to eliminate these errors. It's easier
+to read and visually easier to review for correctness.
+
+## Prior Work
+
+* Interpolated strings have been implemented and well-received in many languages.
+For many such examples, see [String Interpolation](https://en.wikipedia.org/wiki/String_interpolation).
+* Jason Helson has submitted a DIP [String Syntax for Compile-Time Sequences](https://github.com/dlang/DIPs/pull/140).
+* [Adam's string interpolation proposal](http://dpldocs.info/this-week-in-d/Blog.Posted_2019_05_13.html)
+
+## Description
+
+```
+writefln(i"I ate %apples and %{d}bananas totalling %(apples + bananas) fruit.");
+```
+gets rewritten as:
+```
+writefln("I ate %s and %d totalling %s fruit.", apples, bananas, apples + bananas);
+```
+It will also work with printf:
+
+```
+printf(i"I ate %{d}apples and %{d}bananas totalling %{d}(apples + bananas) fruit.\n");
+```
+becomes:
+```
+printf("I ate %s and %d totalling %s fruit.\n", apples, bananas, apples + bananas);
+```
+
+The `{d}` syntax is for when the format specifier needs to be anything other that `s`,
+which is the default. What goes between the `{` `}` is not specified so this capability
+can work with foreseeable format specification improvements without needing to update
+the core language. It also makes interpolated strings agnostic about what the format
+specifications are, as long as they start with `%`.
+
+
+The interpolated string starts as a special string, `InterpolatedString`, which is the same as a
+`DoubleQuotedString` but with an `i` prefix and no `StringPostFix`. This appears in the grammar
+as an `InterpolatedExpression` which is under `PrimaryExpression`.
+
+`InterpolatedExpresssion`s undergo semantic analysis similar to `MixinExpression`.
+The string scanned from left to right, according to the following grammar:
+
+```
+Elements:
+    Element
+    Element Elements
+
+Element:
+    Character
+    '%%'
+    '%' Argument
+    '%' FormatString Argument
+
+FormatString:
+    '{' FormatString '}'
+    CharacterNoBraces
+
+CharacterNoBraces:
+    CharacterNoBrace
+    CharacterNoBrace CharacterNoBraces
+
+CharacterNoBrace:
+    characters excluding '{' and '}'
+
+
+Argument:
+    Identifier
+    Expression
+
+Expression:
+    '(' Expression ')'
+    CharacterNoParens
+
+CharacterNoParens:
+    CharacterNoParen
+    CharacterNoParen CharacterNoParens
+
+CharacterNoParen:
+    characters excluding '(' and ')'
+```
+
+The `InterpolatedExpression` is converted to a tuple expression, where the first element
+is the transformed string literal, and the `Argument`s form the rest of the elements.
+
+The transformed string literal is constructed as follows:
+
+If the `Element` is:
+
+* `Character`, it is written to the output string.
+* `'%%'`, a '%' is written to the output string.
```

Comment:
`'%%'`, a '%' is written to the output string.

Shouldn't it better be that %% stays as a %% in the resulting format string. Or else one will need to put %%%% in the interpolated string to get a % in the result of the writef or the printf.

Reply:
How would you write a `%` to the output otherwise?

Reply:
If during transformation of the interpolated string the %% becomes a %, then the format string will contain an isolated % which is an error in a format string. A double percent in the interpolated string has to stay a double percent in the format string, or else you would have to put 4 % chars.
The transformation is 
interpolated string => format string => output 
Example:
writef(i"Percent %{d}value %%") becomes 
writef("Percent %d %", value)  which is an error (at least undefined behaviour for printf.